### PR TITLE
Add more test cases for cautious-robot

### PR DESCRIPTION
**Pull request by zoeduan**
Added two test cases: test_download_images.py and test_buddycheck.py

### CODE REVIEW

File: tests/test_buddycheck.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        self.buddy_check.buddy_id = 'filename'
+        merged_df = self.buddy_check.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')
+        expected_df = pd.DataFrame({
+            'filename': ['image1.jpg', 'image2.jpg', 'image3.jpg'],
+            'checksum': ['abc123', 'def456', 'ghi789'],
+            'md5': ['abc123', 'def456', 'ghi789']
+        })
+        pd.testing.assert_frame_equal(merged_df, expected_df)
+
+    def test_check_alignment_all_matching(self):
```
Comment (egrace479) [inline tests/test_buddycheck.py:62]: `self.buddy_check.buddy_id = 'filename'` must be set again or you can set another `BuddyCheck` instance with the `buddy_id` defined for those using it (potentially the more efficient method). I'm guessing this will resolve the 6 BuddyCheck tests that are failing.

File: tests/test_buddycheck.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            'MD5': ['abc123', 'def456', 'ghi789']
+        })
+        source_df.columns = source_df.columns.str.lower()
+        checksum_df.columns = checksum_df.columns.str.lower()
+        merged_df = self.buddy_check.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')
+        missing_imgs = self.buddy_check.check_alignment(source_df, merged_df)
+        self.assertIsNone(missing_imgs)
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (egrace479) [inline tests/test_buddycheck.py:163]:
```suggestion
    unittest.main()

```

File: tests/test_download_images.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        mock_process_csv.return_value = mock_data
+
+        with self.assertRaises(SystemExit) as cm:
+            main()
+        
+        self.assertEqual(cm.exception.code, "Exited without executing.")
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (egrace479) [inline tests/test_download_images.py:284]:
```suggestion
    unittest.main()

```

File: tests/test_download_images.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+            "file_url": ["http://test_url1.com/image.jpg", "http://test_url2.com/image.jpg"],
+            "subfolder": ["test_subfolder1", "test_subfolder2"]
+        })
+        self.IMG_DIR = "test_dir"
+        self.LOG_FILEPATH = "test_log_path.jsonl"
+        self.ERROR_LOG_FILEPATH = "test_error_log_path.jsonl"
+        self.DOWNSAMPLE_PATH = "test_downsample_dir"
+        self.DOWNSAMPLE_SIZE = 100
+
+        os.makedirs(self.IMG_DIR, exist_ok=True)
```
Comment (egrace479) [inline tests/test_download_images.py:26]: It can't find the directories for `tests/test_download_images.py`. I'm guessing it's the timing of creation and tear down or that isn't actually happening?

File: tests/test_buddycheck.py (hunk 1 above)
Comment (zoeduan) [reply tests/test_buddycheck.py:62]: Thank you! I have set up another BuddyCheck instance with the buddy_id defined. Now, there are no errors, just a few failed test cases. I will investigate further.

File: tests/test_download_images.py (hunk 4 above)
Comment (zoeduan) [reply tests/test_download_images.py:26]: Thank you! I am guessing it might because the 'downsample_dir' was used before it is defined in the code. I will investigate further.

File: tests/test_buddycheck.py (hunk 1 above)
Comment (zoeduan) [reply tests/test_buddycheck.py:63]: Hello @egrace479. I made these changes: 1. add another BuddyCheck instance with buddy_id initialized. 2. removing the leading whitespace with self.image_source_file and self.check_source_file in setUp. 3. update missing_imgs to retain its original index. Now all the test cases with test_buddycheck run clear.

File: tests/test_download_images.py (hunk 4 above)
Comment (zoeduan) [reply tests/test_download_images.py:26]: Hello @egrace479 The error 'UnboundLocalError for downsample_dir' suggests that downsample_dir is being referenced before it has been assigned a value. Could you please review this? It might be a defect that needs to be addressed. Thank you! 

There are other failed test cases with retry and logging, I will investigate further.

Comment (egrace479) [reply tests/test_download_images.py:26]: Thank you, yes I found the error with the downsampling directory. I set `downsample_dir_path` in the same manner as `image_dir_path` at the beginning of `download_images` (so it's always `downsample_dir_path`, then gets the subfolder addition if subfolders are selected).

File: tests/test_buddycheck.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+import pandas as pd
+import tempfile
+import os
+from cautiousrobot import BuddyCheck
+from cautiousrobot.utils import process_csv
+
+class TestBuddyCheck(unittest.TestCase):
+    def setUp(self):
+        self.buddy_check = BuddyCheck()
+        self.buddy_check_filename = BuddyCheck(buddy_id='filename')
```
Comment (egrace479) [inline tests/test_buddycheck.py:12]:
```suggestion
        self.buddy_check_filename = BuddyCheck(buddy_id='filename')
        self.buddy_check_id_col = BuddyCheck(buddy_id = "filename", buddy_col = "sha256")
```
Add check for the other potential initialization parameters (overwriting defaults)?

File: tests/test_buddycheck.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        self.checksum_source_file.close()
+        
+
+    def tearDown(self):
+        os.remove(self.img_source_file.name)
+        os.remove(self.checksum_source_file.name)
+
+    def test_initialization(self):
+        self.assertEqual(self.buddy_check.buddy_id, None)
+        self.assertEqual(self.buddy_check.buddy_col, 'md5')
```
Comment (egrace479) [inline tests/test_buddycheck.py:38]:
```suggestion
        self.assertEqual(self.buddy_check.buddy_col, 'md5')
        self.assertEqual(self.buddy_check_id_col.buddy_id, 'filename')
        self.assertEqual(self.buddy_check_id_col.buddy_col, 'sha256')
```

File: tests/test_buddycheck.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+    def test_merge_on_checksum(self):
+        source_df = pd.read_csv(self.img_source_file.name)
+        checksum_df = pd.read_csv(self.checksum_source_file.name).drop(columns=['filename'])
```
Comment (egrace479) [inline tests/test_buddycheck.py:42]:
```suggestion
        checksum_df = pd.read_csv(self.checksum_source_file.name)
```
It should work regardless of whether the DataFrame has a "filename" column.

File: tests/test_buddycheck.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        missing_imgs = self.buddy_check_filename.check_alignment(source_df, merged_df)
+        self.assertIsNone(missing_imgs)
+
+    def test_check_alignment_some_missing(self):
+        source_df = pd.DataFrame({
+            'filename': ['image1.jpg', 'image2.jpg', 'image3.jpg', 'image4.jpg'],
+            'checksum': ['abc123', 'def456', 'ghi789', 'jkl012']
+        })
+        checksum_df = pd.read_csv(self.checksum_source_file.name)
+        print(checksum_df)
```
Comment (egrace479) [inline tests/test_buddycheck.py:76]:
```suggestion
```

File: tests/test_buddycheck.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        expected_missing_imgs = pd.DataFrame({
+            'filename': ['image4.jpg'],
+            'checksum': ['jkl012']
+        })
+        pd.testing.assert_frame_equal(missing_imgs.reset_index(drop=True), expected_missing_imgs)
+
+    def test_validate_download_success(self):
+        missing_imgs = self.buddy_check.validate_download(
+            source_df=pd.read_csv(self.img_source_file.name),
+            checksum_df=pd.read_csv(self.checksum_source_file.name).drop(columns=['filename']),
```
Comment (egrace479) [inline tests/test_buddycheck.py:88]:
```suggestion
            checksum_df=pd.read_csv(self.checksum_source_file.name),
```

File: tests/test_buddycheck.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            'md5': ['xyz123', 'uvw456', 'rst789']
+        })
+        merged_df = self.buddy_check_filename.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')
+        missing_imgs = self.buddy_check_filename.check_alignment(source_df, merged_df)
+        self.assertIsNotNone(missing_imgs)
+        self.assertEqual(missing_imgs.shape[0], 3)
+
+    def test_check_alignment_checksums_only(self):
+        source_df = pd.read_csv(self.img_source_file.name)
+        checksum_df = pd.read_csv(self.checksum_source_file.name).drop(columns=['filename'])
```
Comment (egrace479) [inline tests/test_buddycheck.py:125]:
```suggestion
        checksum_df = pd.read_csv(self.checksum_source_file.name)
```

File: tests/test_buddycheck.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        checksum_df = pd.read_csv(self.checksum_source_file.name)
+        merged_df = self.buddy_check.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')
+        missing_imgs = self.buddy_check.check_alignment(source_df, merged_df)
+        self.assertIsNone(missing_imgs)
+
+    def test_check_alignment_empty_checksum_df(self):
+        source_df = pd.read_csv(self.img_source_file.name)
+        checksum_df = pd.DataFrame(columns=['filename', 'md5'])
+        self.buddy_check.buddy_id = 'filename'
+        merged_df = self.buddy_check.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')
```
Comment (egrace479) [inline tests/test_buddycheck.py:141]:
```suggestion
        merged_df = self.buddy_check_filename.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')
```

File: tests/test_buddycheck.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        source_df = pd.DataFrame(columns=['filename', 'checksum'])
+        checksum_df = pd.read_csv(self.checksum_source_file.name)
+        merged_df = self.buddy_check.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')
+        missing_imgs = self.buddy_check.check_alignment(source_df, merged_df)
+        self.assertIsNone(missing_imgs)
+
+    def test_check_alignment_empty_checksum_df(self):
+        source_df = pd.read_csv(self.img_source_file.name)
+        checksum_df = pd.DataFrame(columns=['filename', 'md5'])
+        self.buddy_check.buddy_id = 'filename'
```
Comment (egrace479) [inline tests/test_buddycheck.py:140]:
```suggestion
```
With change below, the definition is unnecessary.

File: tests/test_buddycheck.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        merged_df = self.buddy_check.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')
+        missing_imgs = self.buddy_check.check_alignment(source_df, merged_df)
+        self.assertIsNotNone(missing_imgs)
+        self.assertEqual(missing_imgs.shape[0], 3)
+
+    def test_check_alignment_case_insensitive_columns(self):
```
Comment (egrace479) [inline tests/test_buddycheck.py:146]: Please remove this test; there is no case insensitivity.

File: tests/test_download_images.py [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+        mock_response_success.status_code = 200
+        mock_response_success.raw = BytesIO(b"fake_image_data")
+        get_mock.side_effect = [mock_response_retry] * 2 + [mock_response_success]
+
+        download_images(self.DUMMY_DATA, self.IMG_DIR, self.LOG_FILEPATH, self.ERROR_LOG_FILEPATH, retry=3)
+
+        for filename in self.DUMMY_DATA['filename']:
+            self.assertTrue(os.path.isfile(f"{self.IMG_DIR}/{filename}"))
+
+    @patch('requests.get')
```
Comment (egrace479) [inline tests/test_download_images.py:69]: Can you replace these with a loop over retry codes (I think you could do something similar to @thompsonmj's update on the [sum-buddy `hasher` test](https://github.com/Imageomics/sum-buddy/pull/16/files#diff-d135cf68fe42d10ee2c17c30b1f38af17466b46368aa593f77a8fa45c0e61d73)).

**Review by egrace479 — COMMENTED** [review]
The extra initialization test is up to you, but there were some other areas where it's not quite testing the function (when it's not passed the `buddy_id` column name it should merge on just checksums regardless of whether there is a filename column in the checksum DataFrame).

**Comment by zoeduan** [comment]
Hello @egrace479 @thompsonmj, I have updated the test cases, and now all the test cases run clear. Thanks for review the code!

**Comment by thompsonmj** [comment]
In `test_download_images.py`, I'd add `wait=0.1` to the `test_success_after_retries` and `test_failure_after_retries` tests to make tests go more quickly.

**Comment by thompsonmj** [comment]
I also receive 1 error:
```
FAILED tests/test_buddycheck.py::TestBuddyCheck::test_check_alignment_empty_img_df - ValueError: Unable to fill values because RangeIndex cannot contain NA
```

Details:
```
=================================================================== FAILURES ====================================================================
_______________________________________________ TestBuddyCheck.test_check_alignment_empty_img_df ________________________________________________

self = <test_buddycheck.TestBuddyCheck testMethod=test_check_alignment_empty_img_df>

    def test_check_alignment_empty_img_df(self):
        source_df = pd.DataFrame(columns=['filename', 'checksum'])
        checksum_df = pd.read_csv(self.checksum_source_file.name)
>       merged_df = self.buddy_check.merge_on_filename_checksum(source_df, checksum_df, 'filename', 'checksum')

tests/test_buddycheck.py:136: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/cautiousrobot/buddy_check.py:34: in merge_on_filename_checksum
    merged_df = pd.merge(source_df,
../../.local/lib/python3.8/site-packages/pandas/core/reshape/merge.py:156: in merge
    return op.get_result(copy=copy)
../../.local/lib/python3.8/site-packages/pandas/core/reshape/merge.py:813: in get_result
    self._maybe_add_join_keys(result, left_indexer, right_indexer)
../../.local/lib/python3.8/site-packages/pandas/core/reshape/merge.py:986: in _maybe_add_join_keys
    rvals = algos.take_nd(taker, right_indexer, fill_value=rfill)
../../.local/lib/python3.8/site-packages/pandas/core/array_algos/take.py:110: in take_nd
    return arr.take(
../../.local/lib/python3.8/site-packages/pandas/core/indexes/base.py:1084: in take
    allow_fill = self._maybe_disallow_fill(allow_fill, fill_value, indices)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = RangeIndex(start=0, stop=3, step=1), allow_fill = True, fill_value = 0, indices = array([], dtype=int64)

    @final
    def _maybe_disallow_fill(self, allow_fill: bool, fill_value, indices) -> bool:
        """
        We only use pandas-style take when allow_fill is True _and_
        fill_value is not None.
        """
        if allow_fill and fill_value is not None:
            # only fill if we are passing a non-None fill_value
            if self._can_hold_na:
                if (indices < -1).any():
                    raise ValueError(
                        "When allow_fill=True and fill_value is not None, "
                        "all indices must be >= -1"
                    )
            else:
                cls_name = type(self).__name__
>               raise ValueError(
                    f"Unable to fill values because {cls_name} cannot contain NA"
                )
E               ValueError: Unable to fill values because RangeIndex cannot contain NA

../../.local/lib/python3.8/site-packages/pandas/core/indexes/base.py:1117: ValueError
```

**Comment by zoeduan** [comment]
> In `test_download_images.py`, I'd add `wait=0.1` to the `test_success_after_retries` and `test_failure_after_retries` tests to make tests go more quickly.

Thank you! I updated test_download_images.py, it runs much faster!! Will commit my changes later.

**Comment by egrace479** [comment]
> I also receive 1 error:
> 
> ```
> FAILED tests/test_buddycheck.py::TestBuddyCheck::test_check_alignment_empty_img_df - ValueError: Unable to fill values because RangeIndex cannot contain NA
> ```
> 
> Details:
>  ...

Strange, @thompsonmj, I see everything passing.

I very much like your `wait = .1` suggestion for the retry tests!

**Comment by zoeduan** [comment]
> > I also receive 1 error:
> > ```
> > FAILED tests/test_buddycheck.py::TestBuddyCheck::test_check_alignment_empty_img_df - ValueError: Unable to fill values because RangeIndex cannot contain NA
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Details:
> > ...
> 
> Strange, @thompsonmj, I see everything passing.
> 
> I very much like your `wait = .1` suggestion for the retry tests!

Same with me.

File: tests/testdata/images/image1.jpg
[the remark this answers is not in the document]
Comment (zoeduan) [reply tests/testdata/images/image1.jpg:1]: I downloaded them from google image. If it is inappropriate, we can use other images.

[the remark this answers is not in the document]
Comment (zoeduan) [reply tests/testdata/images/image1.jpg:1]: > What is the source of this (and the other) image?

Thank you! @egrace479 I have updated the testing images.

Comment (egrace479) [inline tests/testdata/images/image1.jpg:1]: Thanks @zoeduan! Could you add the source info (Smithsonian Institution, NMNH & specific department into the commits for both---amend the commit and add a new line with that info for each).

**Comment by thompsonmj** [comment]
On the issue of the `test_check_alignment_empty_img_df ` test, when I execute that test with `pytest`, it raises the `ValueError`, but the test passes when run with `unittest`, which seems odd to me.

Regardless though, since the `merge_on_filename_checksum` method is only called in `validate_download`, we could add an assertion that neither input DataFrame for that method is empty since it would mean the user is doing something that doesn't match their expectations. Then we could remove the tests for empty inputs to `merge_on_filename_checksum` and add a test for `validate_download` with empty inputs, asserting that a custom exception is raised.

How does that sound?

**Review by thompsonmj — APPROVED** [review]

**Review by egrace479 — APPROVED** [review]
Thanks for all the hard work on this one!!
